### PR TITLE
Fixed an issue in min/max of tickets. Fixes #7

### LINKED ISSUE #7: Add a way to keep track of min/max ticket quantities (bundles) and enforce them in the MER cart
Steps to replicate the issue:

1. Set up a ticket bundle or similar ticket that has a min and or max quantity. You could follow either of the examples from these guides:

https://eventespresso.com/wiki/create-ticket-bundle/

https://eventespresso.com/wiki/set-graduated-ticket-pricing/

2. After setting up the event with the  bundle price ( like 10 tickets at a lower price ), then go to the event page and add that bundle to your cart
Screen Shot 2020-02-21 at 1 36 21 PM

3. then go to view the cart,
4. then set the quantity in the cart to 1; You can use either the text input or the + - inputs
Unexpected result:
You get the bundle price but only buy 1 ticket.
Screen Shot 2020-02-21 at 1 37 29 PM

The reverse is also possible where you can add more tickets than the max quantity limit for that ticket.

Brent mentioned that a possible solution would be "if a ticket can have a qty between 5 and 10, then we can just disable the buttons when it hits those limits.

It may also require some validation for the text input fields too.

Reported in these topics:

https://eventespresso.com/topic/minimum-and-maximum-quantity-not-respected-on-cart-page/
https://eventespresso.com/topic/bundle-pricing/#post-234433
https://eventespresso.com/topic/buy-one-get-one-free/#post-256166
https://eventespresso.com/topic/minimum-ticket-bug-update-request/

### PR DESCRIPTION
**Pull request by hoseinrafiei**
I fixed an issue in min / max of tickets so users are not able to order tickets less than bundle minimum or order more than bundle maximum.
Also I fixed another issue on one of the errors regarding reaching maximum tickets per event.

### CODE REVIEW

File: EED_Multi_Event_Registration.module.php [hunk 1]
```diff
@@ -1172,15 +1173,16 @@ function( $additional_info ) {
...
-	 * @param int       $quantity
-	 * @param bool      $cart_update
+	 * @param EE_Ticket    $ticket
+	 * @param int          $quantity
+	 * @param EE_Line_Item $line_item
+	 * @param string       $action
 	 * @return bool
 	 */
-	protected function can_purchase_ticket_quantity( $ticket = null, $quantity = 1, $cart_update = false ) {
+	protected function can_purchase_ticket_quantity( $ticket = null, $quantity = 1, $line_item = null, $action = 'add' ) {
```
Comment (tn3rb) [inline EED_Multi_Event_Registration.module.php]: this is older code so we were not type hinting for objects when this was written, but we do now, so could you please change the above line to:

```suggestion
	protected function can_purchase_ticket_quantity(
		EE_Ticket $ticket = null,
		$quantity = 1,
		EE_Line_Item $line_item = null,
		$action = 'add'
	) {
```

File: EED_Multi_Event_Registration.module.php [hunk 2]
```diff
@@ -1172,15 +1173,16 @@ function( $additional_info ) {
...
 		// any tickets left at all?
 		$tickets_remaining = $ticket instanceof EE_Ticket ? $ticket->remaining() : 0;
-		if ( ! $tickets_remaining ) {
+		if ( ! $tickets_remaining && $action != 'remove') {
```
Comment (tn3rb) [inline EED_Multi_Event_Registration.module.php]: we also try to always use type safe comparison 

```suggestion
		if ( ! $tickets_remaining && $action !== 'remove') {
```

File: EED_Multi_Event_Registration.module.php [hunk 3]
```diff
@@ -1195,7 +1197,7 @@ protected function can_purchase_ticket_quantity( $ticket = null, $quantity = 1,
 		);
 		$limit_error_1 = sprintf( $total_tickets_string, $quantity );
 		// is there enough tickets left to satisfy request?
-		if ( $tickets_remaining < $quantity ) {
+		if ( $tickets_remaining < $quantity && $action != 'remove') {
```
Comment (tn3rb) [inline EED_Multi_Event_Registration.module.php]: same here
```suggestion
		if ( $tickets_remaining < $quantity && $action !== 'remove') {
```

File: EED_Multi_Event_Registration.module.php [hunk 4]
```diff
@@ -1208,9 +1210,53 @@ protected function can_purchase_ticket_quantity( $ticket = null, $quantity = 1,
 			EE_Error::add_error( $limit_error_1 . '<br/>' . $limit_error_2, __FILE__, __FUNCTION__, __LINE__ );
 			return false;
 		}
+
+		// is the quantity allowed based on ticket max?
+		if( $line_item instanceof EE_Line_Item && $ticket instanceof EE_Ticket )
```
Comment (tn3rb) [inline EED_Multi_Event_Registration.module.php]: great 👍

File: EED_Multi_Event_Registration.module.php [hunk 5]
```diff
@@ -1208,9 +1210,53 @@ protected function can_purchase_ticket_quantity( $ticket = null, $quantity = 1,
...
+        {
+            if($action == 'add') $new_quantity = ($line_item->quantity() + $quantity);
+            else if($action == 'remove') $new_quantity = ($line_item->quantity() - $quantity);
+            else $new_quantity = $quantity;
```
Comment (tn3rb) [inline EED_Multi_Event_Registration.module.php]: please also use explicit bracketing for formatting:

```suggestion
            $new_quantity = $quantity;
            if ($action === 'add') {
                $new_quantity = ($line_item->quantity() + $quantity);
            } elseif ($action === 'remove') {
                $new_quantity = ($line_item->quantity() - $quantity);
            }
```

File: EED_Multi_Event_Registration.module.php [hunk 6]
```diff
@@ -1208,9 +1210,53 @@ protected function can_purchase_ticket_quantity( $ticket = null, $quantity = 1,
...
+
+            // can purchase more tickets based on ticket max?
+            $ticket_maximum = $ticket->max();
+            if( $new_quantity > $ticket_maximum )
+            {
+                // translate and possibly pluralize the error
+                $tickets_maximum_string = _n(
```
Comment (tn3rb) [inline EED_Multi_Event_Registration.module.php]:
```suggestion
                $tickets_maximum_string = esc_html__(
```

File: EED_Multi_Event_Registration.module.php [hunk 7]
```diff
@@ -1208,9 +1210,53 @@ protected function can_purchase_ticket_quantity( $ticket = null, $quantity = 1,
...
+                    'The registration limit for this ticket is %1$d ticket per transaction, therefore the total number of tickets you may purchase at this time can not exceed %1$d.',
+                    'The registration limit for this ticket is %1$d tickets per transaction, therefore the total number of tickets you may purchase at this time can not exceed %1$d.',
+                    $ticket_maximum,
+                    'event_espresso'
```
Comment (tn3rb) [inline EED_Multi_Event_Registration.module.php]: great job handling i18n 👏

File: EED_Multi_Event_Registration.module.php [hunk 8]
```diff
@@ -1208,9 +1210,53 @@ protected function can_purchase_ticket_quantity( $ticket = null, $quantity = 1,
...
+                $max_error = sprintf( $tickets_maximum_string, $ticket_maximum );
+                EE_Error::add_error( $max_error, __FILE__, __FUNCTION__, __LINE__ );
+                return false;
+            }
+
+            $ticket_minimum = $ticket->min();
+            if( $ticket_minimum > 0 && ( $new_quantity < $ticket_minimum ) )
+            {
+                // translate and possibly pluralize the error
+                $tickets_minimum_string = _n(
```
Comment (tn3rb) [inline EED_Multi_Event_Registration.module.php]:
```suggestion
                $tickets_minimum_string = esc_html__(
```

File: EED_Multi_Event_Registration.module.php [hunk 9]
```diff
@@ -1208,9 +1210,53 @@ protected function can_purchase_ticket_quantity( $ticket = null, $quantity = 1,
...
+                $min_error = sprintf( $tickets_minimum_string, $ticket_minimum );
+                EE_Error::add_error( $min_error, __FILE__, __FUNCTION__, __LINE__ );
+                return false;
+            }
+        }
+
 		// check event
-		$total_ticket_quantity_within_event_additional_limit = $this->_total_ticket_quantity_within_event_additional_limit( $ticket, $quantity, $cart_update );
-		if ( ! $total_ticket_quantity_within_event_additional_limit ) {
+		$total_ticket_quantity_within_event_additional_limit = $this->_total_ticket_quantity_within_event_additional_limit( $ticket, $quantity, ($action === 'update') );
```
Comment (tn3rb) [inline EED_Multi_Event_Registration.module.php]: 👍

File: EE_Event_Cart_Line_Item_Display_Strategy.php [hunk 10]
```diff
@@ -271,45 +271,55 @@ private function _ticket_row( EE_Line_Item $line_item, $options = array() ) {
...
 	private function _ticket_qty_input( EE_Line_Item $line_item, EE_Ticket $ticket, $required = '' ) {
-		if ( $ticket->remaining() - $line_item->quantity() ) {
-			$disabled = '';
-			$disabled_class = '';
-			$disabled_title = __( 'add one item', 'event_espresso' );
-			$query_args = array( 'event_cart' => 'add_ticket', 'ticket' => $ticket->ID(), 'line_item' => $line_item->code() );
+        $input_disabled_class = '';
+        $input_disabled = '';
+
+		if ( !($ticket->remaining() - $line_item->quantity()) || ($line_item->quantity() >= $ticket->max()) ) {
```
Comment (tn3rb) [inline EE_Event_Cart_Line_Item_Display_Strategy.php]: this is called more than a few times in this method so would be good to extract to a variable like:

```
```suggestion
		$line_item_quantity = $line_item->quantity();
		if ( !($ticket->remaining() - $line_item_quantity) || ($line_item_quantity >= $ticket->max()) ) {
```

same goes with `$line_item->code()` although I know that you didn't touch that in this PR, but making that change as well would be great

File: EE_Event_Cart_Line_Item_Display_Strategy.php [hunk 11]
```diff
@@ -271,45 +271,55 @@ private function _ticket_row( EE_Line_Item $line_item, $options = array() ) {
...
+			<a	title = "' . $remove_disabled_title . '"
+					class="event-cart-remove-ticket-button event-cart-button event-cart-icon-button button' . $required . $remove_disabled_class . '"
 					rel = "' . $line_item->code() . '"
 				    data-target="event-cart-update-txt-qty-' . $line_item->code() . '"
-					href = "' . add_query_arg( array(
-						'event_cart' => 'remove_ticket',
-						'ticket'      => $ticket->ID(),
-						'line_item'   => $line_item->code()
-					), EE_EVENT_QUEUE_BASE_URL ) . '"
+					href = "' . add_query_arg( $remove_query_args, EE_EVENT_QUEUE_BASE_URL ) . '"
```
Comment (tn3rb) [inline EE_Event_Cart_Line_Item_Display_Strategy.php]: improved readability === nice!

**Comment by tn3rb** [comment]
@hoseinrafiei plz refresh the page if some of my code suggestions don't make sense as I had a couple of mistakes in my suggestions that needed to be corrected. Thanks again for the PR.

**Comment by hoseinrafiei** [comment]
Hi Brent,
@tn3rb 

Thanks for your nice feedback.
I reviewed your requested changes and fixed them. Please review the PR again and let me know if there is anything else that should be fixed.

File: EED_Multi_Event_Registration.module.php [hunk 12]
```diff
@@ -1172,62 +1173,116 @@ function( $additional_info ) {
...
+		if ( ! $total_ticket_quantity_within_event_additional_limit && $action != 'remove') {
+		    // New Quantity
+		    if($action == 'update') $quantity = $quantity - $line_item->quantity();
```
Comment (tn3rb) [inline EED_Multi_Event_Registration.module.php]: oops... I missed these during my first review...

- both comparisons need to be type safe
- inner if conditional should use explicit bracketing

File: EE_Event_Cart_Line_Item_Display_Strategy.php [hunk 13]
```diff
@@ -271,45 +271,56 @@ private function _ticket_row( EE_Line_Item $line_item, $options = array() ) {
...
-			$disabled_title = __( 'add one item', 'event_espresso' );
-			$query_args = array( 'event_cart' => 'add_ticket', 'ticket' => $ticket->ID(), 'line_item' => $line_item->code() );
+        $input_disabled_class = '';
+        $input_disabled = '';
+        $line_item_quantity = $line_item->quantity();
+
+		if ( !($ticket->remaining() - $line_item_quantity) || ($line_item_quantity >= $ticket->max()) ) {
+            $input_disabled = ' disabled';
+            $add_disabled_class = $input_disabled_class = ' disabled-event-cart-btn';
+            $add_disabled_title = __( 'there are no more items available', 'event_espresso' );
```
Comment (tn3rb) [inline EE_Event_Cart_Line_Item_Display_Strategy.php]: oops found another translation that needs escaping with `esc_html__()`

File: EE_Event_Cart_Line_Item_Display_Strategy.php [hunk 14]
```diff
@@ -271,45 +271,56 @@ private function _ticket_row( EE_Line_Item $line_item, $options = array() ) {
...
+            $add_query_args = array( 'event_cart' => 'view' );
 		} else {
-			$disabled = ' disabled';
-			$disabled_class = ' disabled-event-cart-btn';
-			$disabled_title = __( 'there are no more items available', 'event_espresso' );
-			$query_args = array( 'event_cart' => 'view' );
+            $add_disabled_class = '';
+            $add_disabled_title = __( 'add one item', 'event_espresso' );
```
Comment (tn3rb) [inline EE_Event_Cart_Line_Item_Display_Strategy.php]: needs escaping with `esc_html__()`

File: EE_Event_Cart_Line_Item_Display_Strategy.php [hunk 15]
```diff
@@ -271,45 +271,56 @@ private function _ticket_row( EE_Line_Item $line_item, $options = array() ) {
...
+            $add_query_args = array( 'event_cart' => 'add_ticket', 'ticket' => $ticket->ID(), 'line_item' => $line_item->code() );
 		}
+
+        if ( ($line_item_quantity <= $ticket->min()) ) {
+            $input_disabled = ' disabled';
+            $remove_disabled_class = $input_disabled_class = ' disabled-event-cart-btn';
+            $remove_disabled_title = __( 'You cannot remove more items', 'event_espresso' );
```
Comment (tn3rb) [inline EE_Event_Cart_Line_Item_Display_Strategy.php]: needs escaping with `esc_html__()`

File: EE_Event_Cart_Line_Item_Display_Strategy.php [hunk 16]
```diff
@@ -271,45 +271,56 @@ private function _ticket_row( EE_Line_Item $line_item, $options = array() ) {
...
+            $remove_query_args = array( 'event_cart' => 'view' );
+        } else {
+            $remove_disabled_class = '';
+            $remove_disabled_title = __( 'remove one item', 'event_espresso' );
```
Comment (tn3rb) [inline EE_Event_Cart_Line_Item_Display_Strategy.php]: needs escaping with `esc_html__()`

**Comment by hoseinrafiei** [comment]
That's fine @tn3rb :) Actually I should find them myself.

I did what you requested and also I replaced all the instances of `__()` with `escape_html__()` in those two files. Please review it again and feel free to let me know if there is anything else.

**Comment by joshfeck** [comment]
QA Tested and ready for merging.

**Pull request merged**